tempodb/encoding/vparquet4/block_traceql.go

```diff
@@ -0,0 +1,2311 @@
+	// is returned.
+	return parquetquery.NewLeftJoinIterator(DefinitionLevelResourceSpans,
+		required, iters, batchCol)
+}
+
+func createServiceStatsIterator(makeIter makeIterFn) parquetquery.Iterator {
+	serviceStatsIters := []parquetquery.Iterator{
+		makeIter(columnPathServiceStatsServiceName, nil, columnPathServiceStatsServiceName),
+		makeIter(columnPathServiceStatsSpanCount, nil, columnPathServiceStatsSpanCount),
+		makeIter(columnPathServiceStatsErrorCount, nil, columnPathServiceStatsErrorCount),
+	}
+	return parquetquery.NewJoinIterator(DefinitionLevelServiceStats, serviceStatsIters, &serviceStatsCollector{})
+}
+
+func createTraceIterator(makeIter makeIterFn, resourceIter parquetquery.Iterator, conds []traceql.Condition, start, end uint64, shardID, shardCount uint32, allConditions bool) (parquetquery.Iterator, error) {
+	traceIters := make([]parquetquery.Iterator, 0, 3)
+
+	var err error
+
+	// add conditional iterators first. this way if someone searches for { traceDuration > 1s && span.foo = "bar"} the query will
+	// be sped up by searching for traceDuration first. note that we can only set the predicates if all conditions is true.
+	// otherwise we just pass the info up to the engine to make a choice
+	for _, cond := range conds {
+		switch cond.Attribute.Intrinsic {
+		case traceql.IntrinsicTraceID:
+			traceIters = append(traceIters, makeIter(columnPathTraceID, NewTraceIDShardingPredicate(shardID, shardCount), columnPathTraceID))
+		case traceql.IntrinsicTraceDuration:
+			var pred parquetquery.Predicate
+			if allConditions {
+				pred, err = createIntPredicate(cond.Op, cond.Operands)
+				if err != nil {
+					return nil, err
+				}
+			}
+			traceIters = append(traceIters, makeIter(columnPathDurationNanos, pred, columnPathDurationNanos))
+		case traceql.IntrinsicTraceStartTime:
+			if start == 0 && end == 0 {
+				traceIters = append(traceIters, makeIter(columnPathStartTimeUnixNano, nil, columnPathStartTimeUnixNano))
+			}
+		case traceql.IntrinsicTraceRootSpan:
+			var pred parquetquery.Predicate
+			if allConditions {
+				pred, err = createStringPredicate(cond.Op, cond.Operands)
+				if err != nil {
+					return nil, err
+				}
+			}
+			traceIters = append(traceIters, makeIter(columnPathRootSpanName, pred, columnPathRootSpanName))
+		case traceql.IntrinsicTraceRootService:
+			var pred parquetquery.Predicate
+			if allConditions {
+				pred, err = createStringPredicate(cond.Op, cond.Operands)
+				if err != nil {
+					return nil, err
+				}
+			}
+			traceIters = append(traceIters, makeIter(columnPathRootServiceName, pred, columnPathRootServiceName))
+		}
+	}
+
+	// order is interesting here. would it be more efficient to grab the span/resource conditions first
+	// or the time range filtering first?
+	traceIters = append(traceIters, resourceIter)
+
+	// collect service stats of the trace
+	traceIters = append(traceIters, createServiceStatsIterator(makeIter))
```

Comment:
we should not add this iterator unless the fetch layer requests it. otherwise we will unnecessarily load this data for every trace before conditions are asserted in the engine

Reply:
I'm happy to work on this, could you give me some tips where to move this iterator?
I noticed it should probably be moved after the time range filtering (not sure why I added it before that), but I guess you've more optimization in mind.

Reply:
the engine builds conditions that it passes into the fetch layer to tell communicate its needs. we currently perform two passes on the data. the first pass asserts any conditions in the query and the second pass requests relevant metadata for the matching rows.

This is the metadata it requests on the second pass.
https://github.com/grafana/tempo/blob/4f53eb9eda1672699898f8ed1d76aac51b530de6/pkg/traceql/storage.go#L15

Currently each condition is requesting a specific column that are span intrinsics. It's likely we don't want to make a new intrinsic here b/c we don't intend for these fields to be refernceable in TraceQL. So we should probably just have a new condition that specifically requests "service metadata".

Then we would only build/add this iterator if that condition were requested.

Reply:
Thanks for the explanation!
I moved the iterator to the second pass: https://github.com/stoewer/tempo/pull/4